src/core/SampleCache.cpp

```diff
@@ -0,0 +1,71 @@
+/*
+ * SampleCache.cpp - Cache for samples
+ *
+ * Copyright (c) 2024 saker <sakertooth@gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SampleCache.h"
+
+#include <QString>
+
+namespace lmms {
+
+SampleCache::Entry::Entry(const QString& filePath, std::shared_ptr<const SampleBuffer> buffer)
+	: m_entry(filePath, buffer)
+{
+}
+
+SampleCache::Entry::~Entry() noexcept
+{
+	auto& [filePath, buffer] = m_entry;
+	(void)filePath;
+	if (buffer.use_count() == 2) { SampleCache::evict(filePath); }
+}
+
+SampleCache::SampleCache()
+{
+	connect(&s_watcher, &QFileSystemWatcher::fileChanged, this, [](const QString& path) { evict(path); });
+}
+
+auto SampleCache::get(const QString& filePath) -> std::optional<Entry>
+{
+	if (filePath.isEmpty()) { return std::nullopt; }
+	if (s_entries.contains(filePath)) { return Entry{filePath, s_entries[filePath].lock()}; };
+
+	try
+	{
+		const auto buffer = std::make_shared<const SampleBuffer>(filePath);
+		if (!s_watcher.addPath(filePath)) { return std::nullopt; }
+		s_entries.insert(filePath, buffer);
+		return Entry{filePath, buffer};
+	}
+	catch (const std::runtime_error& error)
```

Comment:
Is this for catching possible `std::bad_alloc` thrown by `std::make_shared`?

Reply:
It's for catching the failure to create a buffer from a file.

Reply:
I don't see a throw anywhere is why I ask. Perhaps I'm just missing it.

Reply:
D'oh! I thought it was something added with this PR.